sdk/go/common/resource/plugin/plugin_test.go

```diff
@@ -211,3 +211,106 @@ func TestStartupFailure(t *testing.T) {
 	_, err = NewProviderFromPath(ctx.Host, ctx, filepath.Join("testdata", "test-plugin", "test-plugin"))
 	require.ErrorContains(t, err, "could not read plugin [testdata/test-plugin/test-plugin]: not implemented")
 }
+
+func TestNonZeroExitcode(t *testing.T) {
+	d := diagtest.LogSink(t)
+	ctx, err := NewContext(context.Background(), d, d, nil, nil, "", nil, false, nil)
+	require.NoError(t, err)
+
+	pluginPath, err := filepath.Abs("./testdata/provider-language")
+	require.NoError(t, err)
+
+	path := os.Getenv("PATH")
+	t.Setenv("PATH", pluginPath+string(os.PathListSeparator)+path)
+
+	// Check exec.LookPath finds the plugin
+	file, err := exec.LookPath("pulumi-language-test")
+	require.NoError(t, err)
+	require.Contains(t, file, "pulumi-language-test")
+
+	t.Setenv("PULUMI_TEST_PLUGIN_EXITCODE", "1")
+	_, err = NewProviderFromPath(ctx.Host, ctx, filepath.Join("testdata", "test-plugin-exit", "test-plugin-exit"))
+	require.ErrorContains(t, err, "could not read plugin [testdata/test-plugin-exit/test-plugin-exit]: exit status 1")
+
+	// Build a tiny go program that will exit with a non-zero code and run that, check it gives the same result.
+	tmp := t.TempDir()
+	err = os.WriteFile(filepath.Join(tmp, "main.go"), []byte(`
+	package main
+	import "os"
+
+	func main() {
+		os.Exit(1)
+	}
+	`), 0o600)
+	require.NoError(t, err)
+	err = os.WriteFile(filepath.Join(tmp, "go.mod"), []byte(`
+	module test-plugin-exit
+	go 1.24
+	`), 0o600)
+	require.NoError(t, err)
+
+	// Build and run the program
+	cmd := exec.Command("go", "build", "-o", "test-plugin-exit", ".")
+	cmd.Dir = tmp
+	stdout, err := cmd.CombinedOutput()
+	t.Log(string(stdout))
+	require.NoError(t, err)
+
+	_, err = NewProviderFromPath(ctx.Host, ctx, filepath.Join(tmp, "test-plugin-exit"))
+	// the prefix of the error message is unstable because it's in a temp dir but we can check the start and end
+	// separately.
+	require.ErrorContains(t, err, "could not read plugin [")
+	require.ErrorContains(t, err, "test-plugin-exit]: exit status 1")
```

Comment:
Are we expecting the `1` to come from the go program, or from the `PULUMI_TEST_PLUGIN_EXITCODE` env? Would it make sense to change the program to `os.Exit(2)` to differentiate? (I might be misunderstanding the test)